### PR TITLE
Allow transferring points with the consent of those involved

### PR DESCRIPTION
**Pull request by jeffkaufman**
Allow a PR to go in that only transfers points.  Specifically, it must:

* Only have changes in the form "add a file to players/$player/bonuses/"
* Decrease the total number of points
   * Since #33 gives points for merging PRs this can, after counting the point for having
     a PR merged, leave the total number of points unchanged.
* Have been approved by every player losing points.

Examples of running this:

* On #71, which has me giving up a point:

```
./run.sh 71
...
Considering whether this can be merged as a points transfer:
  yes
Meets requirements for points transfer.  PASS
```

* On #70 which adds a file, but not a points file:

```
./run.sh 70
...
Considering whether this can be merged as a points transfer:
Doesn't meet requirements for points transfer: not enough values to unpack (expected 4, got 1)
...
Exception: Insufficient approval
```

* On #59 which was a points transfer that didn't decrease the total number of points but predates named bonuses:

```
./run.sh 59
...
Considering whether this can be merged as a points transfer:
Doesn't meet requirements for points transfer: All file changes must be additions
FYI: this PR has been sitting for 0 days
Approvals: got 5 (jeffkaufman pavellishin shaunagm sligocki tnelling) needed 4

PASS
```

* On #71 which gives me a point:

```
./run.sh 72
...
Considering whether this can be merged as a points transfer:
Doesn't meet requirements for points transfer: points change PRs must on net remove points
FYI: this PR has been sitting for 0 days
Approvals: got 1 (jeffkaufman) needed 4
...
Exception: Insufficient approval
```

### CODE REVIEW

File: validate.py [hunk 1]
```diff
@@ -128,6 +180,10 @@ def determine_if_mergeable():
       else:
         rejections.append(user)
 
+  if mergeable_as_points_transfer(diff, approvals, rejections):
```
Comment (pavellishin) [inline validate.py]: Where's `diff` defined?

**Comment by pavellishin** [comment]
> this will allow people to manufacture points very easily

I'm fine with that, because the points don't matter anyway :)

**Comment by pavellishin** [comment]
I've also noticed a trend - we're also getting a lot of PRs that depend on other PRs. :/ I think some people *cough Jeff cough* have more time to contribute to this than others :P  

I think in the future once we've ironed out some kinks, we'll want to run parallel games at different speeds.

Or maybe once this game is over, we should try to do more of a speed-round - I understand concerns about people who can only check in once or twice a day, but it would be interesting to see how this game plays out if it's quick, and if most people end up not actually reviewing PRs that get merged in until they're already in master.

File: validate.py (hunk 1 above)
Comment (jeffkaufman) [reply validate.py]: In #48

**Comment by jeffkaufman** [comment]
> this will allow people to manufacture points very easily because they will bet a point for submitting a bonus transfer with very small approval threshold.

That's a very good point, and one I wasn't considering.  I'll rewrite it to fix that once #33, #47, and #48 are in.

**Comment by jeffkaufman** [comment]
@pavellishin 
> I think some people cough Jeff cough have more time to contribute to this than others 

Not really.  I should be spending much less time on this.  Hopefully I'll either lose interest somewhat, or have better discipline, but other people taking a long time to review things is really a *good* thing for me, all things considered.

**Comment by sligocki** [comment]
> > this will allow people to manufacture points very easily
> 
> I'm fine with that, because the points don't matter anyway :)

... presumably they matter for something otherwise why are we putting so much effort in to quantify them :P

**Comment by jeffkaufman** [comment]
@pavellishin 

> Is there a simpler way to do this? There's a lot of explicit indices and string comparisons going on. Can we do something conceptually simpler, like checking out master and summing up all the users' points, followed by checking out the current PR and summing all the users' points, and doing arithmetic based on that?

That's possible, and I considered it, but I think we're going to want other parts of validate.py that look at the contents of a PR and make a decision about what conditions to apply.  I think that's much easier if we start building things that look at the diff directly.

The explicit indices and string manipulation are me trying to be paranoid about what counts.  I'm pretty sure the following check would return the same results under all circumstances:

    if (len(path_components) != 4 or
        len(diff_lines) != 8 or
        path_components[0] != 'players' or
        path_components[2] != 'bonuses' or
        diff_lines[7] != ''):

**Comment by jeffkaufman** [comment]
@pavellishin but I can still make this much more readable, and I'll do that when I come back to it.

**Comment by pavellishin** [comment]
Thanks, I would appreciate it!

**Comment by jeffkaufman** [comment]
this is still waiting for #50 before I pick it up again

**Comment by jeffkaufman** [comment]
One thing this needs is figuring out how to have a PR not count for points.  I'm thinking the commit description should start with `nopoints`.

This might be annoying to figure out when inspecting the PR, and then again when inspecting history, we'll see.

**Comment by tnelling** [comment]
Would it make sense for all PRs that transfer points to not themselves grant additional points? Whether that's voluntary self-deduction, or e.g. #59. That way `validate.py` can basically just check to see if `players\bonuses` has been modified. Or maybe even just `players` - I don't know if we want to reward people for removing other players, or given people an extra point when they join.

**Comment by jeffkaufman** [comment]
Updated the PR, rewrote the diff parsing to be more legible, updated description, and added examples to the description.

**Comment by jeffkaufman** [comment]
> this is still waiting for #50 before I pick it up again

this actually didn't need #50 

> One thing this needs is figuring out how to have a PR not count for points. I'm thinking the commit description should start with nopoints.

Yeah, this is too annoying.  But we can get the same effect just by requiring that the total points delta be negative.

**Comment by tnelling** [comment]
This doesn't address all the issues I raised, but does seem like a well-thought-out approach. But can we at least put it into its own file? This is a lot of logic to add to an already bloated file.

**Comment by jeffkaufman** [comment]
@tnelling 
> But can we at least put it into its own file? This is a lot of logic to add to an already bloated file.

I want to think about how to break things into files.  I think that should be a separate PR we can talk about on its own.

(Specifically, I think we can have something where each "if the PR does X then it is/isn't ok to merge" can be in individual files, as a way to reduce merge conflicts)

**Pull request merged**